### PR TITLE
HIVE-25115: Compaction queue entries may accumulate in "ready for cleaning" state

### PR DESCRIPTION
**Pull request by klcopp**
See HIVE-25115

### How was this patch tested?
Unit test

### CODE REVIEW

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 1]
```diff
@@ -305,17 +302,21 @@ public void markCompacted(CompactionInfo info) throws MetaException {
          * By filtering on minOpenTxnWaterMark, we will only cleanup after every transaction is committed, that could see
          * the uncompacted deltas. This way the cleaner can clean up everything that was made obsolete by this compaction.
          */
-        String s = "SELECT \"CQ_ID\", \"CQ_DATABASE\", \"CQ_TABLE\", \"CQ_PARTITION\", "
-                + "\"CQ_TYPE\", \"CQ_RUN_AS\", \"CQ_HIGHEST_WRITE_ID\" FROM \"COMPACTION_QUEUE\" WHERE \"CQ_STATE\" = '"
-                + READY_FOR_CLEANING + "'";
-        if (minOpenTxnWaterMark > 0) {
-          s = s + " AND (\"CQ_NEXT_TXN_ID\" <= " + minOpenTxnWaterMark + " OR \"CQ_NEXT_TXN_ID\" IS NULL)";
-        }
+        StringBuilder sb = new StringBuilder();
```
Comment (deniskuzZ) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:305]: It's really hard to read the query with builder constructs.

Comment (deniskuzZ) [reply standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:305]: I think, with this approach, you might end up not cleaning anything at all as you'll be always looking at latest compaction id that might be blocked by incoming txns.

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 2]
```diff
@@ -305,16 +302,19 @@ public void markCompacted(CompactionInfo info) throws MetaException {
...
+                "   SELECT MAX(\"CQ_NEXT_TXN_ID\") \"CQ_NEXT_TXN_ID\", MAX(\"CQ_ID\") \"CQ_ID\" " +
+                "     FROM \"COMPACTION_QUEUE\" WHERE \"CQ_STATE\" = '" + READY_FOR_CLEANING + "'" +
+                "     AND (\"CQ_NEXT_TXN_ID\" <= "+ minOpenTxnWaterMark + " OR \"CQ_NEXT_TXN_ID\" IS NULL)";
         if (retentionTime > 0) {
-          s = s + " AND \"CQ_COMMIT_TIME\" < (" + getEpochFn(dbProduct) + " - " + retentionTime + ")";
+          s += "      AND \"CQ_COMMIT_TIME\" < (" + getEpochFn(dbProduct) + " - " + retentionTime + ")";
+        }
+        s += "        GROUP BY \"CQ_DATABASE\", \"CQ_TABLE\", \"CQ_PARTITION\" ) \"X\" )";
+        if (LOG.isDebugEnabled()) {
+          LOG.debug("Going to execute query <" + s + ">");
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:316]: `LOG.debug("Going to... <{}>", s);`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java [hunk 3]
```diff
@@ -365,7 +365,29 @@ public void markCleaned(CompactionInfo info) throws MetaException {
...
+          s += " AND \"CQ_PARTITION\" = ?";
+        }
+        s += " AND \"CQ_STATE\"='" + READY_FOR_CLEANING + "' AND \"CQ_ID\" <= " + info.id;
+        pStmt = dbConn.prepareStatement(s);
+        pStmt.setString(1, info.dbname);
+        pStmt.setString(2, info.tableName);
+        if (info.partName != null) {
+          pStmt.setString(3, info.partName);
+        }
+        LOG.debug("Going to execute query <" + s + "> for CQ_ID=" + info.id);
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/CompactionTxnHandler.java:382]: `LOG.debug("Going to execute query <{}> for CQ_ID={}", s, info.id);`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java [hunk 4]
```diff
@@ -383,6 +384,37 @@ public static String getFullTableName(String dbName, String tableName) {
...
+   */
+  public static int executeUpdateQueries(List<String> updateQueries, List<Integer> updateQueryCount, List<Long> ids,
+          Connection dbConn)
+          throws SQLException {
+    int totalCount = 0;
+    int updatedCount = 0;
+    for (int i = 0; i < updateQueries.size(); i++) {
+      String query = updateQueries.get(i);
+      long insertCount = updateQueryCount.get(i);
+      LOG.debug("Going to execute update <" + query + ">");
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java:405]: `LOG.debug("Going to ... <{}>", query);`

File: standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java [hunk 5]
```diff
@@ -383,6 +384,37 @@ public static String getFullTableName(String dbName, String tableName) {
...
+      PreparedStatement pStmt = dbConn.prepareStatement(query);
+      for (int j = 0; j < insertCount; j++) {
+        pStmt.setLong(j + 1, ids.get(totalCount + j));
+      }
+      totalCount += insertCount;
+      int count = pStmt.executeUpdate();
+      LOG.debug("Updated " + count + " records");
```
Comment (belugabehr) [inline standalone-metastore/metastore-server/src/main/java/org/apache/hadoop/hive/metastore/txn/TxnUtils.java:412]: Use anchors.

**Pull request closed without merging**